rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,367 @@
+    repeated string Roles = 1
+        [ (gogoproto.jsontag) = "roles", (gogoproto.casttype) = "SystemRole" ];
+    // JoinMethod is the joining method required in order to use this token.
+    // Supported joining methods include "token", "ec2", and "iam".
+    string JoinMethod = 2
+        [ (gogoproto.jsontag) = "join_method", (gogoproto.casttype) = "JoinMethod" ];
+    // BotName is the name of the bot this token grants access to, if any
+    string BotName = 3 [ (gogoproto.jsontag) = "bot_name,omitempty" ];
+    // SuggestedLabels is a set of labels that resources should set when using this token to enroll
+    // themselves in the cluster
+    wrappers.LabelValues SuggestedLabels = 4 [
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "suggested_labels,omitempty",
+        (gogoproto.customtype) = "Labels"
+    ];
+
+    oneof ProviderConfiguration {
+        ProvisionTokenSpecV3OIDCGoogle OIDCGCP = 5;
+        ProvisionTokenSpecV3AWSIAM IAM = 6;
+        ProvisionTokenSpecV3AWSEC2 EC2 = 7;
+    }
+}
+```
+
+The new schema introduces some duplication of fields between the IAM and EC2
+configuration (e.g `AWSAccount`), but does make it clearer which fields belong
+to EC2 and IAM configuration without the user being reliant on reading the
+documentation of the fields. This has the additional benefit of also reducing
+the complexity of validation code, which is currently reliant on checking
+`JoinMethod` to determine which fields are allowed to be set.
+
+##### Migration
+
+TODO: A strategy for moving between SpecV2 and SpecV3
+
+##### Configuration for OIDC GCP
+
+```yaml
+kind: token
+version: v3
+metadata:
+  name: my-gcp-token
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  roles: [Node]
+  join_method: oidc-gcp
+  gcp:
+    allow:
+    - project_id: "my-project"
+      instance_name: "an-instance"
+    - project_id: "my-project"
+      instance_name: "a-different-instance"
+```
+
+For GCP joining, they will set `join_method` to `oidc-gcp`.
+
+To allow the user to configure rules for what JWTs will be accepted, we will leverage the `allow` field similar to how it works with the IAM joining. Each element of the slice is a set of conditions that must all be satisfied, and a token is authorised as long as at least one allow block is satisfied.
+
+For GCP, we will map the [following JWT claims](https://cloud.google.com/compute/docs/instances/verifying-instance-identity#token_format) to configuration values:
+
+- `sub`: subject
+- `google.compute_engine.project_id`: project_id
+- `google.compute_engine.project_number`: project_number
+- `google.compute_engine.instance_name`: instance_name
+- `google.compute_engine.instance_id`: instance_id
+- `google.compute_engine.zone`: zone
+
+To ensure that we guide users to creating secure configurations, we will also ensure that at least one of the following fields is included in each allow block:
+
+- project_id
+- project_number
+- subject
```

Comment:
You might consider adding `zone` too to only support a list of [high](https://cloud.google.com/security/compliance/fedramp#:~:text=Google%20Cloud%20services%20and%20authorized%20regions%20that%20are%20covered%20by%20FedRAMP) Fedramp compliant workloads' zones?

Reply:
`zone` will indeed be included as an option for a user to configure, but I do not think it belongs in the "at-least-one" mandatory set. The idea of forcing the user to provide at least one key from the mandatory set is to prevent a malicious actor using another google cloud project to generate a JWT that would be used to access a user's Teleport cluster (for example, if the user configured an allow rule with only instance_name, creating a VM in a different project with that name would allow the malicious actor to generate a legitimate JWT matching their allow rule).

Reply:
Sounds good enough then 👌🏼